[BE][Team-05] [coco, nas]반찬 API 서버 기능 구현

Malloc72P opened the pull request:
안녕하세요! 5팀의 백엔드를 맡은 coco와 nas라고 합니다!

이번에는 포스트맨 mock서버를 만들어서 IOS분들에게 제공했습니다
추가적으로 API 설계문서는 Github의 위키에 정리했습니다. [링크](https://github.com/Malloc72P/sidedish/wiki/Sidedish-API-%EC%84%A4%EA%B3%84)

AWS에 스프링 앱서버를 배포했습니다.
이번에는 개발PC에서 빌드하고, JAR파일을 SCP명령어로 AWS서버에 업로드하는 방식으로 배포했습니다
앱서버의 링크는 http://3.34.213.95:8080 입니다

구현한 API는 메인요리, 국물요리, 반찬 목록보기, 상세보기, 주문하기입니다
각각의 상품(음식)에 대한 이미지를 저장하기 위해서 다음과 같이 테이블을 설계했습니다
![image](https://user-images.githubusercontent.com/41663269/115833765-65aadc00-a44f-11eb-8cf1-e2ee4da55de7.png)
상품과 이미지의 다대다 관계를 표현하기 위해서 사이에 ITEM_IMAGE라는 테이블을 만들었습니다

이와 마찬가지로, 이벤트(ex : 초특가 할인 이벤트)를 구현할때도 이미지 테이블과 비슷하게 설계했습니다
![image](https://user-images.githubusercontent.com/41663269/115834126-d3570800-a44f-11eb-9415-37352acf29b5.png)

이상으로 글을 마치겠습니다. 리뷰 잘 부탁드립니다!

Malloc72P committed:
feat : git ignore 제외사항 추가

Malloc72P committed:
feat : 스프링 애플리케이션 생성

Malloc72P committed:
build : DB 연결 설정 추가

Malloc72P committed:
feat : 카테고리와 아이템 엔티티 추가

- 카테고리에서 아이템 목록을 가지는 방향으로 구현
- 해당 기능을 테스트하는 코드 작성

choigiseong committed:
feat : event class , event_item class 작성

- 다대다 관계를 풀어 일대다, 다대일로 구현

choigiseong committed:
test : event item test

- item 조회 시 event item도 조회 되어야 함

choigiseong committed:
feat : getter 추가

- 사용할 수 있는 getter 추가

choigiseong committed:
feat : 카테고리 data insert

- 카테고리 추가

choigiseong committed:
feat : DTO 추가

- item을 조회할 때 DTO를 반환

choigiseong committed:
feat : service 추가

- service 추가

choigiseong committed:
feat : custom exception 추가

- 커스텀 예외 추가

choigiseong committed:
test : item DTO 조회 테스트

- 조회 test

choigiseong committed:
feat : table 추가

- image 테이블 추가

Malloc72P committed:
feat : Item에 이미지목록 추가기능 구현

- SidedishImage는 별개의 어그리게이트이므로 레파지토리 생성함
- 이미지가 썸네일인지 디테일인지 구분하기 위해 ENUM으로 구현함

Malloc72P committed:
feat : 아이템 목록보기 API에서 썸네일도 가져오도록 구현

Malloc72P committed:
refactor : 전체 목록보기용 DTO의 생성자 수정

- 속성값을 받아오기 보다 객체를 받아올 수 있도록 수정

Malloc72P committed:
refactor : 목록보기용 ItemDTO의 이름 수정

- 목록보기용 이므로 간소화된 정보를 제공한다는 의미로 preview를 추가함

choigiseong committed:
feat : controller 추가

- 컨트롤러 추가

choigiseong committed:
Merge pull request #11 from Malloc72P/controller

feat : controller 추가

Malloc72P committed:
feat : 아이템 상세보기 기능 구현

Malloc72P committed:
Merge branch 'backend' into backend-nas

Malloc72P committed:
fix : DTO에 getter없어서 응답해주지 못하던 문제 해결

choigiseong committed:
Merge pull request #13 from Malloc72P/backend-nas

상세보기 서비스 기능 구현

Malloc72P committed:
chore : 안쓰는 코드 제거

Malloc72P committed:
reformat : intelliJ의 리포매팅 기능 수행

Malloc72P committed:
feat : 테스트 데이터 생성하는 코드 작성

choigiseong committed:
feat : 주문 api 추가

- 주문 추가

Malloc72P committed:
fix : 주문하기 API의 결과가 반영되지 않던 오류 해결

- save를 안해서 그랬음.

Malloc72P committed:
feat : exception이 발생했을 때, 발생이유만 클라이언트에게 전송하는 기능 구현

Malloc72P committed:
fix : 누락된 리턴 코드를 추가

Malloc72P committed:
feat : 이미지 못찾은 상황에 대한 예외 처리 추가

Malloc72P committed:
refactor : 패키지 위치 수정

Malloc72P committed:
feat : Response헤더의 content-type에 utf-8 추가

Malloc72P committed:
feat : 이벤트 색상코드를 요구사항에 맞게 수정

Malloc72P committed:
feat : 이벤트에 할인율을 넣고, 이걸로 계산해서 할인가를 구하는 기능 구현

- 해당 기능구현을 위해 Item에서 salePrice를 빼고, event에 sale_rate를 추가함
- 아이템 전체목록보기나 상세보기에선, 해당 아이템의 이벤트가 전부 적용된 최고 할인가를 구해서 출력하도록 구현

ksundong commented on be/sidedish/.gitignore:
```diff
@@ -0,0 +1,264 @@
+### VS Code ###
+.vscode/
+
+# Created by https://www.toptal.com/developers/gitignore/api/windows,intellij,jetbrains,java
```
이전단계에서 말씀드렸던 내용이 반영되었네요. 👍

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/controller/SidedishCategoryController.java:
```diff
@@ -0,0 +1,37 @@
+        this.itemService = itemService;
+    }
+
+    @GetMapping("/{category}")
```
중복이 되는 부분이 있죠?
`@RequestMapping`으로 추출해봅시다.

Malloc72P replied:
수정했습니다!

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/controller/SidedishCategoryController.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    @GetMapping("/{category}")
+    public ResponseEntity<PreviewListDtoWrapper> previewItemList(@PathVariable String category) {
```
왜 `ResponseEntity`로 리턴하는 방식을 사용해주셨나요?
장점은 무엇인가요?

ksundong replied:
`Wrapper`라는 네이밍은 별 의미가 없을 것 같은데 그냥 DTO면 안될까요?
이미 DTO 자체가 가진 의미가 Data Transfer Object여서, Wrapping한다는 느낌을 충분히 주는 것 같습니다.

choigiseong replied:
RestController를 사용중.
 별다른 추가 정보가 없을 때는 `@ResponseStatus`만 붙여서 반환해도 된다고 생각합니다.
 `ResponseEntity` 는 더 추가 정보를 담을 수 있다고 알고 있습니다. (컨텐츠 타입,길이)
지금은  `ResponseStatus`로 충분할 것 같지만 멋있어서 사용했습니다.

Malloc72P replied:
Wrapper는 다른 DTO랑 구분하기 위해서 붙였습니다. 너무 햇갈리더라구요.
`SidedishDetailItemDTO`를 감싸주는 DTO가 필요했어요. breaking change를 회피하기 위해서 json의 형태를 아래와 같이 했습니다.
```json
{
   item : {
     /*...*/
   }
}
```
해당 Response에 맞는 클래스의 이름을 뭐로 지어줄지가 막막했습니다. 고민하다가 결국 Wrapper로 지었습니다

ksundong replied:
ResponseEntity는 제한적으로 사용하시길 권장드립니다. 대부분은 Annotation으로 충분히 정리 가능합니다.

---

객체 형태의 API가 Breaking Change가 발생하려면 어떤 상황이 발생해야할까요?
"item"으로 정해준다 하더라도, 같은 상황을 방지할 수 있을까요?
감싸주는 이유에 대해서 잘 생각해보시길 바랍니다 ㅎㅎ

Malloc72P replied:
저는 Breaking Change를 Response의 변경이 클라이언트 코드에 영향을 주는거라고 생각했어요.
그렇다면 새로운 데이터를 보내주는거라면 Braking Change는 발생하지 않을 것 같아요. 그냥 새로운 키를 만들고, 거기에 값을 넣어주면 되니까요.
다만, 기존에 쓰던 키가 수정되면 Breaking Change가 발생할 것 같아요. 근데 이건 프론트와 백엔드간의 약속이 바뀐거니까 어쩔 수 없지 않나 라는 생각이 들어요.
그래서 저는 감싸주는 형태의 DTO면 요구사항에 유연하게 대응할 수 있을 것 같았는데, 꼭 그렇지는 않은 모양이군요. 흠... 어렵네요... 코코랑 좀 더 공부해보겠습니다!

ksundong replied:
네, 감싸주는 형태가 되면
item.abcd.efg 같은 형태가 될 수 있습니다.
item이 계속해서 반복되는 형태는 좋지 않다고 생각해요.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/controller/SidedishCategoryController.java:
```diff
@@ -0,0 +1,37 @@
+    @PostMapping("/{category}/{id}")
+    public void orderItem(@PathVariable String category, @PathVariable Long id, @RequestBody OrderDTO orderDTO) {
+        itemService.order(category, id, orderDTO);
+    }
```
좀더 RESTful 한 API를 디자인 했다면, 리턴 타입은 상관없지만 적어도 ResponseStatus는 201 CREATED를 주는 것이 좋겠네요.

Malloc72P replied:
200 ok외에 다른 상태코드가 있다는걸 처음 알았습니다!
바로 적용해보겠습니다!

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/controller/SidedishCategoryExceptionController.java:
```diff
@@ -0,0 +1,25 @@
+@ControllerAdvice
+public class SidedishCategoryExceptionController {
+
+    @ExceptionHandler({CategoryNotFoundException.class, ItemNotFoundException.class, EventNotFoundException.class})
```
자 개선의견 드려볼게요.
NotFoundException을 만들고, 이를 상속하는 구조로 만들면 어떨 것 같나요?

Malloc72P replied:
그러면 익셉션 핸들러에서 RuntimeException대신 NotFoundException으로 받아서 처리할 수 있겠네요!

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/controller/SidedishCategoryExceptionController.java:
```diff
@@ -0,0 +1,25 @@
+public class SidedishCategoryExceptionController {
+
+    @ExceptionHandler({CategoryNotFoundException.class, ItemNotFoundException.class, EventNotFoundException.class})
+    public ResponseEntity<ErrorResponseDTO> handleNotFoundException(RuntimeException exception) {
```
RuntimeException이 아니라 NotFoundException이라면 더 세세한 예외 처리가 가능해질 수도 있습니다.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/controller/SidedishCategoryExceptionController.java:
```diff
@@ -0,0 +1,25 @@
+
+    @ExceptionHandler({CategoryNotFoundException.class, ItemNotFoundException.class, EventNotFoundException.class})
+    public ResponseEntity<ErrorResponseDTO> handleNotFoundException(RuntimeException exception) {
+        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(new ErrorResponseDTO(exception.getMessage()));
```
이건 ResponseEntity를 쓰지않고도 충분히 가능해요.

Malloc72P replied:
ResponseEntity가 멋있어보여서 이렇게 했던건데, 지금보니까 오히려 가독성만 해치고 있었네요. 어노테이션으로 하는게 좀 더 보기 좋을 것 같습니다

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/controller/SidedishCategoryExceptionController.java:
```diff
@@ -0,0 +1,25 @@
+
+    @ExceptionHandler(OutOfStockException.class)
+    public ResponseEntity<ErrorResponseDTO> handleOutOfStockException(OutOfStockException exception) {
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(new ErrorResponseDTO(exception.getMessage()));
```
OutOfStock이 400에러인가요?
사용자가 잘못 요청했다고는 보여지지 않는데요. 
204 No Content의 사용이 더 적절하지 않을까요? 저도 고민되는 항목이네요~

ksundong replied:
https://www.wolfgangdigital.com/blog/how-to-manage-out-of-stock-products-for-e-commerce-seo
한 번 읽어보세요!

Malloc72P replied:
우선 주문이 실패한거니까 4xx시리즈의 응답코드를 사용해야 한다고 생각했어요. 근데 마땅히 Out of Order를 표현하는 코드를 못찾겠더라구요. 
재고가 없는데 주문하겠다고 요청한거니까 잘못 요청한게 아닐까? 라고 생각해서 BadRequest를 주었습니다.

ksundong replied:
정답이 없는 문제기는 한데, 사용자가 재고 있는 상황에서 다른 사용자가 재고를 소진해서 발생하는 경우도 있지 않을까요?

Malloc72P replied:
그러게요... 어렵네요.... 이게 HTTP 응답코드를 가지고 비즈니스 상황을 표현하려다보니 곤란해지는 것 같아요... 이 부분도 좀 더 고민해보겠습니다!

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/SidedishCategory.java:
```diff
@@ -0,0 +1,54 @@
+    private String categoryName;
+    private List<SidedishItem> sidedishItemList = new ArrayList<>();
+
+    public SidedishCategory(String categoryName) {
+        this.categoryName = categoryName;
+    }
+
+    protected SidedishCategory() {
+    }
```
nit: 되도록이면 기본 생성자가 위에 위치하는것이 좋을 것 같군요.

choigiseong replied:
수정 하였습니다!

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/SidedishCategory.java:
```diff
@@ -0,0 +1,54 @@
+        return sidedishItemList;
+    }
+
+    public SidedishItem findItem(Long itemId){
```
```suggestion
    public SidedishItem findItem(Long itemId) {
```

ksundong replied:
이런 메서드 정의 좋네요.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/SidedishCategory.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public SidedishItem findItem(Long itemId){
+        return sidedishItemList.stream()
+                .filter(item -> item.isSameId(itemId))
+                .findFirst()
+                .orElseThrow(ItemNotFoundException::new);
```
id가 아닌 index를 받으면 O(n)이 아니라 O(1)이 가능하겠네요.

choigiseong replied:
저희가 order와 showDetailItem 메소드에서 findItem를 사용하고 있습니다.
둘의 공통점은 클라이언트에서 id값을 통해 해당 item을 얻어온다는 점입니다.
디온께서 말씀하신 index 부분을 생각해 봤는데  저희가 index를 클라이언트에 보내줘야 하고, 클라이언트가 index정보로 아이템 조회를 한다 라고 말씀하신 게 맞을까요?

ksundong replied:
네 그런식으로 하는 편이 더 좋지 않을까 하는 생각이에요.

Malloc72P replied:
흠... 초기에 설계를 좀 더 꼼꼼하게 했으면 좋았을텐데 라는 아쉬움이 남네요 ㅠㅠ 지금 저걸 바꾸게되면 프론트까지 변경의 영향을 받게 되어서 쉽게 수정하기 힘들 것 같습니다. 우선 Oauth 로그인을 구현하고, 인덱스를 사용할 수 있도록 하겠습니다!

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/SidedishItem.java:
```diff
@@ -0,0 +1,158 @@
+
+    private List<SidedishItemImage> sidedishItemImages = new ArrayList<>();
+
+    public SidedishItem(String itemName, String itemDescription, int itemNormalPrice,
+                        int itemQuantity, int itemPointRate, String itemDeliveryInfo, String itemDeliveryFee) {
+        this.itemName = itemName;
+        this.itemDescription = itemDescription;
+        this.itemNormalPrice = itemNormalPrice;
+        this.itemQuantity = itemQuantity;
+        this.itemPointRate = itemPointRate;
+        this.itemDeliveryInfo = itemDeliveryInfo;
+        this.itemDeliveryFee = itemDeliveryFee;
+    }
```
빌더패턴을 적용해보는 것도 좋겠네요.
https://johngrib.github.io/wiki/builder-pattern/

Malloc72P replied:
여유가 생기면 적용해보겠습니다
지금은 OAuth가 너무 무서워서 빌더패턴까지 가능할지 잘 모르겠습니다
최대한 노력해보겠습니다

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/SidedishItem.java:
```diff
@@ -0,0 +1,158 @@
+    protected SidedishItem() {
+    }
+
+    public int calculateSalePrice(Set<SidedishEvent> sidedishEvents) {
```
sideDishEvents를 외부에서 받는 이유가 뭔가요?

Malloc72P replied:
SidedishItem에는 EventItem에 대한 Set은 있지만 Event정보는 없습니다.
할인된 가격을 계산하려면 EventItem에 있는 외래키를 이용해서 Event객체를 획득해야 합니다. 그러다보니 외부에서 SidedishEvents를 받아서 할인가격을 계산하는 방식으로 구현했습니다.

ksundong replied:
다대다 매핑인 것 같은데, 저는 안받는 방향이 어떤가하는 생각이 드네요.
그래프를 타서 충분히 알 수 있을 것 같다는 생각이 들어요.

choigiseong replied:
그래프를 탄다는 게 SidedishItem이 가지고 있는 필드 Set<SidedishEventItem> eventItems를 사용하라는 말씀이신가요?
 SidedishEventItem에는 Long id와 외래키들만 가지고 있는데 여기서 어떻게 Event를 얻어 올 수 있나요?
다대다를 일대다, 다대일로 풀었을 때 중간 객체는  객체를 가지고 있음 안되는 것으로 알고 있습니다.

생각해본 바로는 SidedishEventItem에서 리포지토리를 받는 방법이 있을 거 같은데 이거는 최악의 설계인 거 같습니다.
아니면 다른 방법이 있을까요?

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/SidedishItem.java:
```diff
@@ -0,0 +1,158 @@
+    public int calculateSalePrice(Set<SidedishEvent> sidedishEvents) {
+        int salePrice = itemNormalPrice;
+        for (SidedishEvent currentEvent : sidedishEvents) {
+            salePrice -= currentEvent.salePrice(this);
```
여기 로직이 조금 이상해요. event의 할인가격을 this를 받아서 한다면, 서로 사이클이 생기는거 아닌가요?
둘 다 서로를 아는 형태는 좋지 않다고 봐요. 둘 중하나가 포함되는 관계 같은데요?

Malloc72P replied:
이 부분은 저희가 메서드이름을 잘못지어서 오해를 만든 것 같습니다.
`currentEvent.salePrice(this)`는 해당 이벤트를 적용했을 때, 정상가에서 얼마만큼 빼줘야 하는지를 구하는 메서드입니다. 그런데 할인가를 표현하는 단어도 salePrice이고, 얼마만큼 빼줄지를 구하는 메서드도 salePrice이다보니 로직이 이상해보이는 것 같습니다.
그래서 `currentEvent.salePrice(this)`를 `currentEvent.discount(this)`로 이름을 수정했습니다.

ksundong replied:
음... Event가 SidedishItem에 포함되는 관계를 의미한 거였는데, salePrice가 discount로 바뀌어도 다를게 없다고 생각이 되네요.
제가 코드를 완전히 이해를 하지 않아서 그럴 수도 있겠네요.

choigiseong replied:
해결방법을 생각해 봤습니다. 
1. Event의 discount가 int normalPrice를 받아서 할인할 가격을 return해준다.
2. discount메소드를 item쪽으로 넘겨서 사용을 한다.
이렇게 하면 사이클은 깨질 거 같습니다.

처음에 Event에서 item을 받는 이유는 확장성이 있다고 생각을 했었습니다.
정작 이것 때문에 실수를 하는 순간 순환참조 발생해 프로그램이 죽는 문제는 깊게 생각해 보지 못했습니다.
디온 덕분에 하나 더 알아가는 거 같습니다. 감사합니다!

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/SidedishItem.java:
```diff
@@ -0,0 +1,158 @@
+        if (salePrice < 0) {
+            salePrice = 0;
+        }
+        return salePrice;
```
그래도 도메인을 적극적으로 활용하신 부분 좋습니다.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/SidedishItem.java:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    public boolean order(OrderDTO orderDTO) {
+        if (!isPurchasable(orderDTO)) {
```
orderDTO를 넘겨서 물어보기보다는 Order에게 물어보는 형태로 구현은 어때요?

choigiseong replied:
order라는 entity는 따로 가지고 있지 않고 저희는 주문을 할 때 json body로 quantity를 받아옵니다. 
그 요청을 받을 때 생성되는 게 orderDTO입니다.
즉 orderDTO는 주문 수량만 가지고 있는 형태가 됩니다.
 그 주문 수량으로 item의 재고수량을 체크하는 것 입니다.

order에게 물어보라는 말씀은 orderEntity를 만들어 설계하라는 말씀이신가요?

ksundong replied:
Order Entity가 될 수도 있고, OrderDTO에게 물어봐도 괜찮겠죠.
private 메서드의 생성은 피하는 편이 좋습니다.

ksundong replied:
아 코드를 다시읽으니 어떤 느낌으로 작성되었는지 알겠네요.
어떻게 하면 협력을 표현할 수 있을까요?

Malloc72P replied:
`협력을 표현한다`에 대해 코코와 얘기해봤습니다
구매가능한지 물어보는건 ItemService이고, 이 요청을 받아서 구매가능여부를 알려주는건 SidedishItem이라고 생각합니다.
이때, Item한테 요청메세지를 보내는데, OrderDTO를 보내는게 과연 적절한가? 라는 생각이 들었습니다.
OrderDTO의 책임은 뭘까? 얘의 책임은 프론트로부터 받은 Json데이터를 객체로 파싱하는 것 이라고 생각합니다.
그런데 이걸 isPurchasable의 아규먼트로 보내는건 책임이 모호해지는게 아닌가 라는 생각이 듭니다.

또, OrderDTO가 없지만, 일정 수량만큼 구매가 가능한지 Item한테 물어보는 상황이 생긴다면, 위의 메서드로 처리할 수 없을 것으로 보입니다. 그러면 차라리 isPurchasable의 매개변수는 `int purchseQuantity`로 처리하는게 좋을 것 같습니다. 그러면 협력을 좀 더 잘 표현하는게 아닐까 싶습니다! 디온은 어떻게 생각하시나요?

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/SidedishItem.java:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    public boolean isPurchasable(OrderDTO orderDTO) {
+        return itemQuantity - orderDTO.getQuantity() >= 0;
```
```suggestion
        return itemQuantity >= orderDTO.getQuantity();
```

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/dto/DetailItemDtoWrapper.java:
```diff
@@ -0,0 +1,14 @@
+package com.codesquad.sidedish.category.domain.dto;
+
+public class DetailItemDtoWrapper {
+
+    private SidedishDetailItemDTO item;
+
+    public DetailItemDtoWrapper(SidedishDetailItemDTO item) {
+        this.item = item;
+    }
+
+    public SidedishDetailItemDTO getItem() {
+        return item;
+    }
```
왜 한 번 더 감싸주신건가요?

Malloc72P replied:
breaking change를 회피하기 위해서 API의 Response를 아래와 같이  설계했습니다.
```
{
   item : {  }
}
```
그러다보니 위와 같은 형식의 DTO가 필요해서 만들었습니다.
문제는 이름을 뭘로 지어줘야 하는지가 애매했습니다.
DTO를 감싸주는거니까 Wrapper가 좋겠다는 생각이 들었습니다.
그래서 한번 더 감싸주게 되었습니다.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/dto/OrderDTO.java:
```diff
@@ -0,0 +1,16 @@
+        this.quantity = quantity;
+    }
+
+    protected OrderDTO() {
+    }
```
이건 왜 만들어주신거에요??

choigiseong replied:
spring boot에서 json 요청이 들어올 때 jackson2httpmessageconverter를 통해 객체를 만든다고 알고 있습니다.
내부 로직에서는 objectMapper를 사용하는데 그 때 기본 생성자가 필요하다고 알고있습니다.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/dto/PreviewListDtoWrapper.java:
```diff
@@ -0,0 +1,22 @@
+
+    public PreviewListDtoWrapper(List<SidedishItemPreviewDTO> previewDTOs) {
+        this.items = new ArrayList<>();
+        addDTO(previewDTOs);
```
굳이 메서드로 빼줘야 했을까요?

Malloc72P replied:
한번 해보고 싶었습니다 ㅎㅎ

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/dto/SidedishDetailItemDTO.java:
```diff
@@ -0,0 +1,109 @@
+        this.DeliveryFee = sidedishItem.getItemDeliveryFee();
+
+        this.detailImages = detailImages.stream().map(SidedishImage::getImageUrl).collect(Collectors.toList());
+        this.descriptionImages = descriptionImages.stream().map(SidedishImage::getImageUrl).collect(Collectors.toList());
```
생성자 밖에서 해주는건 어떤가요?

Malloc72P replied:
서비스에서 처리해서 넘겨주도록 수정했습니다!

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/dto/SidedishDetailItemDTO.java:
```diff
@@ -0,0 +1,109 @@
+                ", DeliveryInfo='" + DeliveryInfo + '\n' +
+                ", DeliveryFee='" + DeliveryFee + '\n' +
+                '}';
+    }
```
Object 클래스의 메서드를 오버라이딩 한 것은 맨 밑에 위치시키는게 컨벤션입니다.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/exception/CategoryNotFoundException.java:
```diff
@@ -0,0 +1,9 @@
+package com.codesquad.sidedish.category.exception;
+
+public class CategoryNotFoundException extends RuntimeException {
+    private static final String ERROR_REASON = "존재하지 않는 카테고리 입니다";
```
ERROR_REASON 보다 명확한 네이밍을 사용해주세요.

Malloc72P replied:
`CategoryNotFoundException.ERROR_REASON`이런 식으로 되는거니까, `ERROR_REASON`이면 충분하지 않을까 라고 생각해서 이렇게 했습니다.
혹시 더 좋은 이름이 있을까요?

ksundong replied:
ERROR_REASON이 모두 다 중복되는 이름 아닌가요? 에러메시지를 표현하는 형태면 더 좋을 것 같네요.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/service/SidedishItemService.java:
```diff
@@ -0,0 +1,126 @@
+import java.util.stream.Collectors;
+
+@Service
+@Transactional(readOnly = true)
```
👍

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/service/SidedishItemService.java:
```diff
@@ -0,0 +1,126 @@
+            SidedishImage thumbnailImage = findThumbnailImage(item);
+            itemDTOs.add(new SidedishItemPreviewDTO(item, salePrice, eventDtoSet, thumbnailImage));
+        }
+        return new PreviewListDtoWrapper(itemDTOs);
```
코드가 눈에 잘 안들어오네요.
stream을 이용해서 개선해볼까요?

choigiseong replied:
`SidedishItemService`에 있던 책임을 다른 서비스로 이전을 해봤습니다. 
그래서 아래와 같이 변경을 했습니다!
```

    public PreviewListDtoWrapper showPreviewItems(String categoryName) {
        SidedishCategory category = findByCategoryName(categoryName);
        List<SidedishItem> items = category.getSidedishItemList();

        List<SidedishItemPreviewDTO> itemDTOs = new ArrayList<>();
        for (SidedishItem item : items) {
            Set<SidedishEvent> eventSet = eventFinder.findEvents(item);
            SidedishImage thumbnailImage = imageFinder.findThumbnailImage(item);
            itemDTOs.add(new SidedishItemPreviewDTO(item, eventSet, thumbnailImage));
        }
        return new PreviewListDtoWrapper(itemDTOs);
    }
```

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/service/SidedishItemService.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    public DetailItemDtoWrapper showItem(String categoryName, Long itemId) {
+        SidedishCategory sidedishCategory = sidedishCategoryRepository.findByCategoryName(categoryName).orElseThrow(CategoryNotFoundException::new);
```
동일한 로직이 나오죠? 메서드로 추출해봅시다.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/service/SidedishItemService.java:
```diff
@@ -0,0 +1,126 @@
+
+    public DetailItemDtoWrapper showItem(String categoryName, Long itemId) {
+        SidedishCategory sidedishCategory = sidedishCategoryRepository.findByCategoryName(categoryName).orElseThrow(CategoryNotFoundException::new);
+        SidedishItem sidedishItem = sidedishCategory.findItem(itemId);
```
item의 index를 받았다면 어땠을까요?

Malloc72P replied:
index로 받는다면 List에서 더 효율적으로 검색이 가능할 것 같아요!

저 한가지 질문이 있습니다.
지금 로직은 카테고리를 레파지토리에서 꺼내오고, 꺼낸 카테고리를 통해서 아이템에 접근하는 구조입니다. 이렇게 한 이유는 카테고리가 어그리게이트 루트라고 생각했기 때문입니다. 루트를 통해 접근해야 한다고 생각해서 이렇게 한건데, 다시 생각해보니, 꼭 이렇게 해야 할까라는 생각이 들었습니다.
그냥 아이템 레파지토리를 만들고, 이걸로 아이템의 ID를 써서 찾으면 인덱스를 타고 검색하는거니까 검색효율도 좋아질 것 같고, 또 불필요한 다른 정보(카테고리정보, 찾으려는 아이템 외의 정보)는 찾아오지 않아서 좋을 것 같습니다. 이렇게 해도 될까요? 혹시 문제가 생길 수 있을까요?

ksundong replied:
어려운 문제군요 ㅎㅎ
둘 다 해보시는게 방법이라고 생각됩니다. 둘 중에 좋은걸 취사선택 하는 것이 좋다고 봅니다.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/service/SidedishItemService.java:
```diff
@@ -0,0 +1,126 @@
+        SidedishItem sidedishItem = sidedishCategory.findItem(itemId);
+
+        Set<SidedishEvent> eventSet = createEventSet(sidedishItem);
+        Set<SidedishEventDTO> eventDtoSet = eventSetToDtoSet(eventSet);
```
별로 OOP 적이지 않네요.

Malloc72P replied:
EventFinder랑 ImageFinder라는 서비스를 만들고, 위의 메서드로 하던 책임을 위임하는 방식으로 고쳐보았고, 이들한테 요청하는식으로 수정했습니다.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/service/SidedishItemService.java:
```diff
@@ -0,0 +1,126 @@
+        return new DetailItemDtoWrapper(sidedishDetailItemDTO);
+    }
+
+    private SidedishImage findThumbnailImage(SidedishItem item) {
```
제가 생각하기엔 private 메서드들이 다른 클래스에 위치하는 형태는 어떤지 질문드리고 싶네요.

Malloc72P replied:
ImageFinder라는 서비스를 만들고, 해당 기능에 관한 책임을 위임하는 방식으로 수정했습니다.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/service/SidedishItemService.java:
```diff
@@ -0,0 +1,126 @@
+            return sidedishImageRepository.findById(sidedishItemImage.getSidedishImage()).orElseThrow(ImageNotFoundException::new);
+        } catch (ImageItemNotFoundException e) {
+            return DefaultImageUtil.getNotFoundImage();
+        }
```
try catch 없이 풀어내는 형태가 좋을 거 같은데요?

ksundong replied:
orElseGet을 사용해봅시다.

Malloc72P replied:
ImageItem은 Item과 Image사이의 관계를 표현하는 객체입니다.
만약 여기서 ImageItem을 못찾았다면, 디폴트이미지를 보여주고자 했습니다. 그런데 디폴트이미지는 어떠한 상품에도 소속되지 않아서 ImageItem의 객체로 만들어줄 수 없었습니다. 대신 디폴트 이미지를 SidedishImage객체로 만들었습니다. 그러다보니, orElseGet()을 쓸 수 없게 되었습니다.

또한, ImageItem을 못찾았을때의 예외는 Checked Exception으로 만들었습니다. 반드시 예외처리를 하도록 강제해서, 디폴트 이미지를 반환할 수 있게 만들고 싶었습니다.

혹시 더 좋은 방법이 있을까요? 저도 이런 이유로 저렇게 하긴 했지만, 확실히 로직이 너무 복잡하다고 느껴서요. 어렵네요...

ksundong replied:
SidedishItemImage sidedishItemImage
이게 문제 아닌가요? 추상화를 이용해서 해결할 수 있을 것 같네요.

Malloc72P replied:
음.. SidedishItemImage랑 SidedishImage를 어떻게 추상화할지가 감이 잘 안오네요. 조금만 더 힌트를 주실 수 있을까요?

저희가 생각하는 추상화는 공통된 속성을 일반화해서 상위클래스를 만드는 것으로 이해했습니다. 그런데 SidedishItemImage와 SidedishImage는 서로 공통된 속성이 없다고 생각합니다.  그래서 추상화를 어떻게 적용해봐야 할지가 감이 잘 안오네요...

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/service/SidedishItemService.java:
```diff
@@ -0,0 +1,126 @@
+        return sidedishImageRepository.findAllById(itemImageList);
+    }
+
+    private Set<SidedishEvent> createEventSet(SidedishItem item, Map<Long, SidedishEvent> eventMap) {
+        for (SidedishEventItem eventItem : item.getEventItems()) {
+            Long eventId = eventItem.getSidedishEvent();
+            if (!eventMap.containsKey(eventId)) {
+                eventMap.put(eventId, sidedishEventRepository.findById(eventId).orElseThrow(EventNotFoundException::new));
+            }
+        }
+        return item.getEventItems().stream()
+                .map(eventItem -> eventMap.get(eventItem.getSidedishEvent()))
+                .collect(Collectors.toSet());
+    }
+
+    private Set<SidedishEvent> createEventSet(SidedishItem item) {
+        Set<SidedishEvent> eventSet = new HashSet<>();
+        for (SidedishEventItem eventItem : item.getEventItems()) {
+            Long eventId = eventItem.getSidedishEvent();
+            SidedishEvent sidedishEvent = sidedishEventRepository.findById(eventId).orElseThrow(EventNotFoundException::new);
+            eventSet.add(sidedishEvent);
+        }
+        return eventSet;
```
코드가 복잡한 느낌이 많이드네요.

choigiseong replied:
처음 설계는 item들을 가지고 올 때 Map으로 event를 관리하게 했습니다.
item들 중 같은 event가 있을 경우 또 DB에서 가지고 오게 하고 싶지 않았기 때문입니다.

하지만 위와 같은 이유 때문에 코드의 복잡도가 높아지는 것 같아 Map로직을 삭제하고 item마다 event를 DB에서 다시 가져오도록 변경했습니다.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/service/SidedishItemService.java:
```diff
@@ -0,0 +1,126 @@
+    @Transactional
+    public void order(String categoryName, Long id, OrderDTO orderDTO) {
+        SidedishCategory category = sidedishCategoryRepository.findByCategoryName(categoryName).orElseThrow(CategoryNotFoundException::new);
+        SidedishItem item = category.findItem(id);
```
굳이 이걸 꺼내서 order를 해야했을까요?

Malloc72P replied:
사실 이 부분에 대해서 많은 고민을 했습니다.
`그냥 아이템의 아이디만 가지고 찾아도 되지 않을까`랑 `어그리게이트 루트를 통해서 찾아야 되지 않을까` 이 두가지 방법을 두고 고민했습니다.
결국 `어그리게이트 루트를 통해서 찾자`로 결정해서 이렇게 했는데, 어떻게 하는게 좋을까요? 저희도 굳이 카테고리를 통해서 findItem하는 것 보다 그냥 아이템 레파지토리를 만들고 이걸로 `findById`해서 아이템을 꺼낸다음 order하는게 더 효율적일 것 같은데, 괜찮을까요?

ksundong replied:
일단 해보시는걸 추천합니다. 제가 DDD에 익숙치않아서, 비효율적인 코드가 생성되는 부분은 걱정이 되네요.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/event/domain/SidedishEvent.java:
```diff
@@ -0,0 +1,59 @@
+        this.eventSaleRate = eventSaleRate;
+    }
+
+    public int salePrice(SidedishItem item){
```
```suggestion
    public int salePrice(SidedishItem item) {
```

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/event/domain/SidedishEvent.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public int salePrice(SidedishItem item){
+        return (int) (item.getItemNormalPrice() * (eventSaleRate / 100));
```
좋네요~

ksundong replied:
get을 안쓰는 방향으로도 개선해보세요~

Malloc72P replied:
음 저희는 객체로 넘기는 편이 좋다고 생각해서 이렇게 구현했습니다.
혹시 이런 상황에서는 속성을 넘기는 편이 더 좋을까요?
getter를 안쓰려면 Item 객체를 직접 넘기지 않고 속성을 넘기는 방법 외에는 없을 것 같아서요. 어떤게 좋을지 잘 판단이 안서네요.

ksundong replied:
getter를 쓰지말라는 의미를 잘 이해하지 못하신 것 같아요.
객체지향적으로 풀어봅시다.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/image/controller/SidedishImageExceptionController.java:
```diff
@@ -0,0 +1,17 @@
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
+public class SidedishImageExceptionController {
```
개별 ControllerAdvice를 정의해주신 이유가 있나요?
그러면 패키지를 한정해줘야 하는건 아닐까요?

ksundong replied:
그리고 RestControllerAdvice도 있는데, 둘의 차이도 한 번 알아보시면 좋겠네요.

Malloc72P replied:
GlobalControllerAdvice를 만들어서 모든 예외에 대한 핸들러를 처리하도록 수정했습니다.
또, RestControllerAdvice를 찾아보니, `ControllerAdvice` + `ResponseBody`를 합쳐놓았다는걸 알게 되었습니다. 이전에는 ResponseEntity가 ResponseBody의 역할을 해주었던게 아닌가 싶네요.
ResponseEntity를 걷어내고 RestControllerAdvice를 사용하도록 수정했습니다.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/image/exception/ImageItemNotFoundException.java:
```diff
@@ -0,0 +1,6 @@
+package com.codesquad.sidedish.image.exception;
+
+public class ImageItemNotFoundException extends Exception {
```
왜 그냥 Exception을 상속하나요?

Malloc72P replied:
`ImageItemNotFoundException` 예외는 반드시 try-catch로 감싸주어서, 이미지를 못찾았을때의 예외처리를 강제하기 위함입니다.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/util/DefaultImageUtil.java:
```diff
@@ -0,0 +1,19 @@
+import com.codesquad.sidedish.image.domain.SidedishImage;
+
+public class DefaultImageUtil {
+    public static final String NOT_FOUND_IMAGE_URL = "https://grepp-programmers.s3.ap-northeast-2.amazonaws.com/files/production/db00837d-4762-4073-abaf-aa924e507eaa/emoticon-01.png";
```
이건 왜 있는건가요?

Malloc72P replied:
이미지를 못찾았을때 디폴트 이미지를 리턴하고자 했습니다. 그때 사용할 용도로 만들어두었습니다.

ksundong commented on be/sidedish/src/main/resources/data.sql:
```diff
@@ -0,0 +1,5 @@
+USE SIDEDISH;
```
파일마다 쿼리문의 Case가 다 다르네요.
일관성을 지켜주세요.

Malloc72P replied:
대문자를 쓰는 방향으로 수정했습니다!

ksundong commented on be/sidedish/src/test/java/com/codesquad/sidedish/SidedishApplicationTests.java:
```diff
@@ -0,0 +1,13 @@
+
+@SpringBootTest
+@Transactional
+class SidedishApplicationTests {
```
필요없으면 제거해주셔도 됩니다.

ksundong reviewed (changes requested):
안녕하세요! Coco & NAS! 리뷰어 Dion입니다.
벌써 두번째 프로젝트네요. 시간이 참 빠른 것 같습니다.
아직 일주일 정도밖에 안되었는데, 많은 부분을 개발해주셨던 것 같아요.
또한, 부족하지만 OOP적인 설계를 하시려고 하는 부분이 눈에 띄어서 좋았습니다.

배포 부분은, 정말 다양한 방법이 있을 수 있겠지만, 수동이 아니라 자동을 지향해보세요~

- 아쉬웠던 점은 Service에서 너무 복잡하게 진행되는 느낌을 받았습니다. 데이터를 가져오는 것은 가져오는 것 대로, 변환하는 것은 변환하는 것 대로 분리를 해주면 가독성이 더 좋아지지 않을까 싶네요.
- ResponseEntity를 사용해야하는 명백한 이유가 없다면 사용을 권하고 싶지는 않습니다.

고생하셨습니다. 몇 몇 개선의견 남겨드리니 확인해주세요~
궁금하거나, 도움이 필요한 부분, 이해가 되지 않는 부분, 토론하고 싶은 부분은 언제든 코멘트 남겨주세요~

choigiseong committed:
refactor : 서비스의 책임 분리

- 각 어그리게이트 별로 서비스를 만들어 분리 했습니다.

choigiseong committed:
refactor : controller에서 동일한 url 매핑

choigiseong committed:
refactor : order의 상태코드를 201 로 변경

- 상태 코드 변경

choigiseong committed:
refactor : NotFoundException을 정의

- 상위 타입으로 NotFoundException을 정의했습니다.

choigiseong committed:
refactor : responseEntity를 responseStatus로 변경

- responseStatus로 충분한다

choigiseong committed:
style : 기본 생성자 위치 변경

- 기본 생성자를 위로

choigiseong committed:
style : 기본 생성자를 위로 이동

choigiseong committed:
style : 기본생성자를 위로

choigiseong committed:
refactor : salePrice 메소드명 discount로 변경

- 할인된 수치 이기에 이름이 적절하기 않다고 판단해 변경

choigiseong committed:
refactor :  isPurchasable 공식 변경

choigiseong committed:
style : toString은 맨 밑에 위치

choigiseong committed:
refactor : 생성자에서의 loop 제거

- service에서 모든 url을 만들어서 주입

choigiseong committed:
refactor : ExceptionController를 글로벌로 배치

- 어그리게이트 별로 존재 한다고 생각했지만 다른 어그리게이트의 예외도 잡고 있었다
- 그래서 각자 만들려고 했지만 글로벌로 변경했다.

choigiseong committed:
style : 코드 포맷

- 대문자로 변경

choigiseong committed:
test : test 삭제

- 테스트 삭제

choigiseong committed:
feat : 세일중이 아닐때는 salePrice는 -1

- ios와 정하기를 세일 중이 아니면 salePrice는 -1을 보내기로 했다.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/dto/OrderDTO.java:
```diff
@@ -0,0 +1,16 @@
+package com.codesquad.sidedish.category.domain.dto;
+
+public class OrderDTO {
+    private int quantity;
```
nit: 어디는 필드와 클래스 선언부가 붙어있고, 떨어져있고 한데요, 일관성있는 코드가 좋을 것 같습니다.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/dto/SidedishDetailItemDTO.java:
```diff
@@ -0,0 +1,109 @@
+
+public class SidedishDetailItemDTO {
+
+    private Long id;
+    private List<String> detailImages;
+    private List<String> descriptionImages;
+    private String name;
+    private String description;
+    private int normalPrice;
+    private int salePrice;
+    private Set<SidedishEventDTO> eventBadgeList;
+    private int PointRate;
+    private boolean isPurchasable;
+    private String DeliveryInfo;
+    private String DeliveryFee;
```
필드가 너무 많다는 생각이 듭니다.
적절하게 객체로 분리하는 것도 방법이라고 생각되네요.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/service/SidedishItemService.java:
```diff
@@ -0,0 +1,77 @@
+public class SidedishItemService {
+
+    private final SidedishCategoryRepository sidedishCategoryRepository;
+    private final ImageFinder imageFinder;
+    private final EventFinder eventFinder;
```
뭔가 새로운 시도같네요. 👍

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/service/SidedishItemService.java:
```diff
@@ -0,0 +1,77 @@
+
+        List<SidedishItemPreviewDTO> itemDTOs = new ArrayList<>();
+        for (SidedishItem item : items) {
+            Set<SidedishEvent> eventSet = eventFinder.findEvents(item);
+            SidedishImage thumbnailImage = imageFinder.findThumbnailImage(item);
```
매 번 조회하는 형태라면 느려질 가능성도 염두에 두는게 좋을 것 같아요.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/event/domain/SidedishEvent.java:
```diff
@@ -0,0 +1,62 @@
+
+public class SidedishEvent {
+
+    public static final int NOT_ON_SALE = -1;
```
이건 여기서 쓰이지 않는데 왜 여기 선언되어 있을까요?
그리고 이런류의 상태를 나타내는 값은 Enum이 더 적절할 것 같네요~

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/event/domain/dto/SidedishEventDTO.java:
```diff
@@ -0,0 +1,45 @@
+        this.colorHex = sidedishEvent.getEventColor();
+    }
+
+    public static Set<SidedishEventDTO> eventSetToDtoSet(Set<SidedishEvent> eventSet) {
+        return eventSet.stream().map(SidedishEventDTO::new)
+                .collect(Collectors.toSet());
+    }
```
자바 컨벤션상 static 메서드는 instance 메서드보다 윗 부분에 위치해야합니다.
이 메서드가 여기 정의되는 것도 조금 고민되는 부분이긴 하네요~

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/event/service/EventFinder.java:
```diff
@@ -0,0 +1,34 @@
+import java.util.stream.Collectors;
+
+@Service
+public class EventFinder {
```
이런식으로 Service 네이밍 하는거 좋네요.
가능하면 `@Transactional(readonly = true)` 까지 해줬으면 더 좋았을 것 같네요.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/event/service/EventFinder.java:
```diff
@@ -0,0 +1,34 @@
+        Set<SidedishEventItem> eventItems = item.getEventItems();
+        return eventItems.stream()
+                .map(SidedishEventItem::getSidedishEvent)
+                .map(id -> findById(id))
```
```suggestion
                .map(this::findById)
```

ksundong replied:
매번 쿼리 날리게 될텐데... 걱정이 되네요 ㅎㅎ

ksundong replied:
차라리 쿼리를 직접 정의해서 In을 이용해보는건 어떨까요?

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/global/controller/GlobalExceptionController.java:
```diff
@@ -0,0 +1,34 @@
+@RestControllerAdvice
+public class GlobalExceptionController {
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    @ExceptionHandler({CategoryNotFoundException.class, ItemNotFoundException.class, EventNotFoundException.class})
```
```suggestion
    @ExceptionHandler(NotFoundException.class)
```

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/global/controller/GlobalExceptionController.java:
```diff
@@ -0,0 +1,34 @@
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    @ExceptionHandler({CategoryNotFoundException.class, ItemNotFoundException.class, EventNotFoundException.class})
+    public ErrorResponseDTO handleNotFoundException(NotFoundException exception) {
+        return new ErrorResponseDTO(exception.getMessage());
+    }
+
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    @ExceptionHandler(OutOfStockException.class)
+    public ErrorResponseDTO handleOutOfStockException(OutOfStockException exception) {
+        return new ErrorResponseDTO(exception.getMessage());
+    }
+
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    @ExceptionHandler(ImageNotFoundException.class)
+    public ErrorResponseDTO handleImageNotFoundException(ImageNotFoundException exception) {
+        return new ErrorResponseDTO(exception.getMessage());
+    }
```
처리가 다 같다는 점이 아쉽네요.
다 같다면 그냥 공통의 예외로 처리해주는 편이 나았을 지도 모르겠습니다.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/global/exception/NotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package com.codesquad.sidedish.global.exception;
+
+public class NotFoundException extends RuntimeException {
+    public NotFoundException(String message) {
+        super(message);
+    }
```
기본 생성자도 만들어주면 좋을 것 같네요~

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/image/domain/SidedishImage.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    public static List<String> imageListToImageUrlList(List<SidedishImage> imageList) {
+        return imageList.stream().map(SidedishImage::getImageUrl).collect(Collectors.toList());
```
개행을 통해서 가독성을 높여주는 방식을 추천드리겠습니다.

ksundong replied:
`static` 메서드의 사용도 고민을 많이 해주세요.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/image/service/ImageFinder.java:
```diff
@@ -0,0 +1,43 @@
+        this.sidedishImageRepository = sidedishImageRepository;
+    }
+
+    public SidedishImage findThumbnailImage(SidedishItem item) {
+        try {
+            //todo : 예외 처리
+            SidedishItemImage sidedishItemImage = item.findThumbnailImage();
+            return findImageById(sidedishItemImage.getId());
+        } catch (ImageItemNotFoundException e) {
+            return DefaultImageUtil.getNotFoundImage();
+        }
+    }
```
개선이 되었다 하더라도 아쉬운게 사실이네요.
`return findImageById(item.findThumbnailImage.getId()).orElseGet(() → DefaultImageUtil.getNotFoundImage);`

이런 형태는 안되나요?
위의 코드도 많이 아쉬운게 사실이긴하네요.

ksundong commented on be/sidedish/src/main/resources/application.properties:
```diff
@@ -0,0 +1,15 @@
+# log
+logging.level.com.codesquad.=DEBUG
```
```suggestion
logging.level.com.codesquad=DEBUG
```

ksundong reviewed (approved):
피드백 빠르게 반영해주시느라 고생많으셨습니다.
서비스를 다른 이름으로 지어주는 것도 좋은 것 같아요. 무엇을 할 건지 책임이 확실하게 와닿네요.
고생하셨습니다~
자잘한 피드백 또 남겨보겠습니다. 화이팅하세요~

choigiseong committed:
refactor : 순환참조 예방

choigiseong committed:
refactor : 파리미터 타입 변화

choigiseong committed:
style : 오타 수정

choigiseong committed:
feat : oauth 구현

- access_token 얻어오기

Malloc72P committed:
feat : Github에서 구글 OAuth 로 변경

- 구글로 변경되면서 Callback 컨트롤러에서 추가로 받을 매개변수를 추가함
- 구글 API를 통해서 userProfile 정보를 가져오는 기능 추가

Malloc72P committed:
feat : gitIgnore에서 누락되어야 하는 파일 추가

Malloc72P committed:
feat : 누락시킬 파일을 gitIgnore에 추가

Malloc72P committed:
feat : json파일로부터 읽어온 정보를 가지고 API요청하도록 구현

- git으로부터 ignore된 server.secret.json파일로부터 cliend id와 secret정보를 읽어오도록 구현
- 앱 초기설정으로 해야 하므로 커맨드라인 러너에 추가함
- 추가로 api url등을 관리하는 상수 클래스를 생성

Malloc72P committed:
fix : 잘못된 final키워드 추가로 기능이 마비되는 이슈 해결

- 엔티티는 변할 수 있는 객체라고 생각해서 다시 final키워드를 제거함

Malloc72P committed:
feat : OAuth API를 요청하는 서비스 생성

Malloc72P committed:
feat : oauth 로그인했을 때 처리하는 로직 추가

- 유저정보가 없으면 생성하고, 있으면 토큰을 업데이트하도록 구현

Malloc72P committed:
feat : jwt 토큰 생성을 위한 의존성 추가

Malloc72P committed:
feat : JWT토큰 생성기능 구현

- 아직 검증기능은 불완전하므로, 보완하는 기능 추가가 필요함

ksundong commented:
머지를 하려고 했는데 변경사항이 좀 많네요.
conflict 해결해주세요~

ksundong commented on .gitignore:
```diff
@@ -0,0 +1 @@
+be/sidedish/src/main/resources/server.secret.json
```
백엔드 쪽에 있는게 낫지 않았을까요?

ksundong commented on be/sidedish/build.gradle:
```diff
@@ -0,0 +1,26 @@
+	runtimeOnly 'mysql:mysql-connector-java'
+	developmentOnly 'org.springframework.boot:spring-boot-devtools'
+	testImplementation 'org.springframework.boot:spring-boot-starter-test'
+	implementation 'io.jsonwebtoken:jjwt:0.9.1'
```
저는 써보니까 jjwt보단 java-jwt가 더 낫더라고요 ㅎㅎ

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/controller/SidedishCategoryController.java:
```diff
@@ -0,0 +1,54 @@
+        }else {
+            System.out.println("not ok");
+            return;
+        }
```
JWT 관련 로직이 있어선 안될 곳에 들어있는 느낌이네요.

ksundong replied:
또 `HttpServletRequest` 를 받지 않고 `@RequestHeader` 애노테이션을 활용해보는건 어떨까 싶네요.
Tokenizer의 사용은 좋네요 ㅎㅎ
System.out은.. 음.. 못본걸로 할까요?

choigiseong replied:
해당 기능들은 인터셉터로 이동시켰습니다!

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/SidedishCategory.java:
```diff
@@ -0,0 +1,54 @@
+    private Long id;
+
+    private String categoryName;
+    private  List<SidedishItem> sidedishItemList = new ArrayList<>();
```
```suggestion
    private List<SidedishItem> sidedishItemList = new ArrayList<>();
```

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/user/controller/UserController.java:
```diff
@@ -0,0 +1,49 @@
+        HttpEntity<RequestAccessTokenDTO> httpEntity = new HttpEntity<RequestAccessTokenDTO>(requestAccessTokenDTO,httpHeaders);
+        String url = "https://github.com/login/oauth/access_token";
+        String body = template.exchange(url, HttpMethod.POST, httpEntity, String.class).getBody();
+        System.out.println(body);
```
별도의 서비스 Layer에 요청을 해도 좋을 것 같네요~
중간커밋이라 정돈되지 않은 내용들이 많이 보이는 것 같아요.

choigiseong replied:
`GoogleApiRequester`를 정의해서 사용하도록 했습니다!

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/SidedishItem.java:
```diff
@@ -0,0 +1,175 @@
+    public int calculateSalePrice(Set<SidedishEvent> sidedishEvents) {
+        int salePrice = itemNormalPrice;
+
+        if (sidedishEvents.size() == 0) {
+            return SidedishEvent.NOT_ON_SALE;
+        }
+
+        for (SidedishEvent currentEvent : sidedishEvents) {
+            salePrice -= currentEvent.discount(itemNormalPrice);
+        }
+        if (salePrice < 0) {
+            salePrice = 0;
+        }
```
0이라는 리터럴들을 상수로 꺼내봅시다.

choigiseong replied:
수정했습니다!

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/category/domain/dto/SidedishDetailItemDTO.java:
```diff
@@ -0,0 +1,109 @@
+
+public class SidedishDetailItemDTO {
+
+    private final Long id;
+    private final List<String> detailImages;
+    private final List<String> descriptionImages;
+    private final String name;
+    private final String description;
+    private final int normalPrice;
+    private final int salePrice;
+    private final Set<SidedishEventDTO> eventBadgeList;
+    private final int PointRate;
+    private final boolean isPurchasable;
+    private final String DeliveryInfo;
+    private final String DeliveryFee;
```
굳이 DTO에 final을 적용할 필요는 없습니다 ㅎㅎ

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/event/domain/SidedishEvent.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    public int discount(int normalPrice) {
+        return (int) (normalPrice * (eventSaleRate / 100));
```
여기도 리터럴 제거해주세요~

choigiseong replied:
수정했습니다!

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/init/InitDataCommandLineRunner.java:
```diff
@@ -0,0 +1,158 @@
+import java.util.List;
+
+@Component
+public class InitDataCommandLineRunner implements CommandLineRunner {
```
저는 Data를 파싱해서 DB에 넣었던 기억이 납니다. ㅎㅎ

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/user/controller/UserController.java:
```diff
@@ -0,0 +1,44 @@
+@RequestMapping("/users")
+public class UserController {
+
+    private final GoogleApiRequester googleApiRequester;
```
네이밍 좋네요.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/user/controller/UserController.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    @GetMapping("/callback")
+    public ResponseJwtTokenDTO oauthCallBack(@RequestParam("code") String code, @RequestParam("scope") String scope,
+                                             @RequestParam("authuser") String authuser, @RequestParam("prompt") String prompt) throws JsonProcessingException {
```
DTO 만들어서 받아보는건 어때요?

choigiseong replied:
`AuthorizationInfo` 라는 DTO를 만들어 받도록 변경했습니다!

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/user/controller/UserController.java:
```diff
@@ -0,0 +1,44 @@
+
+    @GetMapping("/login")
+    public void login(HttpServletResponse response) throws IOException {
+        response.sendRedirect(LOGIN_PAGE_URL);
```
이미 JwtToken이 있는 경우에도 다시 발급하나요?

Malloc72P replied:
이미 있는 경우엔 리프래시 토큰을 발급해야 하지만, 아직 거기까진 못했습니다.
좀 더 공부하고 토큰을 리프래시 할 수 있도록 공부해봐야겠습니다.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/user/controller/UserController.java:
```diff
@@ -0,0 +1,44 @@
+        ReceiveAccessTokenDTO receiveAccessTokenDTO = googleApiRequester.requestAccessToken(code);
+        UserInfoDTO userInfoDTO = googleApiRequester.requestUserInfo(receiveAccessTokenDTO.getAccess_token());
+        String jwtToken = userService.login(userInfoDTO);
+        return new ResponseJwtTokenDTO(jwtToken);
```
```suggestion
        return new JwtTokenResponse(jwtToken);
```

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/user/controller/UserController.java:
```diff
@@ -0,0 +1,44 @@
+    public ResponseJwtTokenDTO oauthCallBack(@RequestParam("code") String code, @RequestParam("scope") String scope,
+                                             @RequestParam("authuser") String authuser, @RequestParam("prompt") String prompt) throws JsonProcessingException {
+        ReceiveAccessTokenDTO receiveAccessTokenDTO = googleApiRequester.requestAccessToken(code);
+        UserInfoDTO userInfoDTO = googleApiRequester.requestUserInfo(receiveAccessTokenDTO.getAccess_token());
```
```suggestion
        UserInfoDTO userInfoDTO = googleApiRequester.requestUserInfo(receiveAccessTokenDTO.getAccessToken());
```

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/user/controller/UserController.java:
```diff
@@ -0,0 +1,44 @@
+    @GetMapping("/callback")
+    public ResponseJwtTokenDTO oauthCallBack(@RequestParam("code") String code, @RequestParam("scope") String scope,
+                                             @RequestParam("authuser") String authuser, @RequestParam("prompt") String prompt) throws JsonProcessingException {
+        ReceiveAccessTokenDTO receiveAccessTokenDTO = googleApiRequester.requestAccessToken(code);
```
```suggestion
        AccessTokenDTO AccessTokenDTO = googleApiRequester.requestAccessToken(code);
```
이래도 충분할 것 같네요~

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/user/domain/User.java:
```diff
@@ -0,0 +1,95 @@
+import org.springframework.data.annotation.Id;
+import org.springframework.data.relational.core.mapping.Table;
+
+@Table("USERS")
```
👍

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/user/domain/User.java:
```diff
@@ -0,0 +1,95 @@
+    protected User() {
+    }
+
+    public User(UserInfoDTO userInfoDTO) {
+        userId = userInfoDTO.getId();
+        email = userInfoDTO.getEmail();
+        verifiedEmail = userInfoDTO.getVerified_email();
+        name = userInfoDTO.getName();
+        familyName = userInfoDTO.getFamily_name();
+        givenName = userInfoDTO.getGiven_name();
+        picture = userInfoDTO.getPicture();
+        locale = userInfoDTO.getLocale();
+    }
```
도메인의 생성자에서 DTO에 의존하는 형태는 지양해주세요.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/user/domain/User.java:
```diff
@@ -0,0 +1,95 @@
+        locale = userInfoDTO.getLocale();
+    }
+
+    public void update(UserInfoDTO userInfoDTO) {
+        userId = userInfoDTO.getId();
+        email = userInfoDTO.getEmail();
+        verifiedEmail = userInfoDTO.getVerified_email();
+        name = userInfoDTO.getName();
+        familyName = userInfoDTO.getFamily_name();
+        givenName = userInfoDTO.getGiven_name();
+        picture = userInfoDTO.getPicture();
+        locale = userInfoDTO.getLocale();
+    }
```
별도의 업데이트용 DTO를 만드는 편이 더 좋아보입니다.

ksundong replied:
혹은 DTO를 Domain객체로 변환 후에 Update를 수행해주는 것도 고려해봐주세요.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/user/dto/ReceiveAccessTokenDTO.java:
```diff
@@ -0,0 +1,44 @@
+
+public class ReceiveAccessTokenDTO {
+
+    private String access_token;
+    private int expires_in;
+    private String scope;
+    private String token_type;
+    private String id_token;
```
네이밍 컨벤션을 지켜주세요.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/user/dto/RequestAccessTokenDTO.java:
```diff
@@ -0,0 +1,43 @@
+
+    private static final String GRANT_TYPE = "authorization_code";
+
+    private String client_id;
+    private String client_secret;
+    private String code;
+    private String grant_type;
+    private String redirect_uri;
```
여기도 마찬가지입니다.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/user/dto/ResponseJwtTokenDTO.java:
```diff
@@ -0,0 +1,16 @@
+public class ResponseJwtTokenDTO {
+    private String token;
+
+    protected ResponseJwtTokenDTO() {
+    }
```
어디는 protected고 어디는 public인데 기준이 무엇인가요?

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/user/service/GoogleApiRequester.java:
```diff
@@ -0,0 +1,44 @@
+
+    private final RestTemplate restTemplate;
+
+    public GoogleApiRequester(RestTemplateBuilder templateBuilder) {
+        this.restTemplate = templateBuilder.build();
+    }
```
스프링 Bean으로 이미 등록되어있을 텐데 이렇게 해주신 이유가 무엇인가요?

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/user/service/UserService.java:
```diff
@@ -0,0 +1,46 @@
+        this.userRepository = userRepository;
+    }
+
+    public String login(UserInfoDTO userInfoDTO) {
+        Optional<User> optionalUser = findUserByUserId(userInfoDTO.getId());
+        User user;
+        if (optionalUser.isPresent()) {
+            user = optionalUser.get();
+            user.update(userInfoDTO);
+            userRepository.save(user);
+        } else {
+            user = createuser(userInfoDTO);
+        }
+        String jwtToken = JwtUtil.createToken(user);
+        logger.info("jwtToken : {}", jwtToken);
+        return jwtToken;
+    }
```
여기 로직은 개선이 필요해보이네요.
다음 리뷰어 분에게 리뷰를 맡기겠습니다 ㅎㅎ

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/util/JwtUtil.java:
```diff
@@ -0,0 +1,42 @@
+    private JwtUtil() {
+    }
+
+    public static void initServerSecretKey() {
+        serverSecretKey = Base64.getEncoder().encodeToString(SecretUtil.serverSecret().getBytes(StandardCharsets.UTF_8));
+    }
```
왜 이렇게 구현해주셨나요?

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/util/JwtUtil.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.Date;
+
+public class JwtUtil {
+    private static String serverSecretKey;
```
static 필드를 어디서든 변경할 수 있는 형태는 지양해주세요.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/util/JwtUtil.java:
```diff
@@ -0,0 +1,42 @@
+
+public class JwtUtil {
+    private static String serverSecretKey;
+    private static final String issuer = "sidedish";
+    private static final long tokenValidTime = 6 * 60 * 60 * 1000L;
```
네이밍 컨벤션 지켜주세요~

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/util/JwtUtil.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.Base64;
+import java.util.Date;
+
+public class JwtUtil {
```
Service로 구현하는것도 생각해봐주세요~

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/util/SecretUtil.java:
```diff
@@ -0,0 +1,40 @@
+            InputStream inputStream = new FileInputStream(serverSecretFile);
+            ObjectMapper objectMapper = new ObjectMapper();
+            SERVER_SECRET_DTO = objectMapper.readValue(inputStream, ServerSecretDTO.class);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
```
무책임한 형태의 예외처리입니다. 지양해주세요.

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/util/UrlConstant.java:
```diff
@@ -0,0 +1,19 @@
+
+import static com.codesquad.sidedish.util.SecretUtil.clientId;
+
+public class UrlConstant {
```
Enum을 적용해봐도 좋겠죠?

ksundong commented on be/sidedish/src/main/java/com/codesquad/sidedish/user/exception/UserNotFoundException.java:
```diff
@@ -0,0 +1,10 @@
+import com.codesquad.sidedish.global.exception.NotFoundException;
+
+public class UserNotFoundException extends NotFoundException {
+    private static final String ERROR_MESSAGE = "해당하는 유저 정보를 찾을 수 없습니다";
+    public UserNotFoundException() {
```
```suggestion
    private static final String ERROR_MESSAGE = "해당하는 유저 정보를 찾을 수 없습니다";
    
    public UserNotFoundException() {
```

ksundong replied:
입력된 유저정보를 변수로 받아서 처리해주는 것도 고려해볼만 할 것 같아요.

ksundong reviewed (approved):
추가로 리뷰해보았습니다.
컨플릭트 해결해주시면 머지하도록 할게요~
고생하셨습니다~

Malloc72P commented:
으악 ㅋㅋ 디온 죄송합니다! ㅠㅠ 저희가 브랜치를 새로 파서 작업해야 했던건데, 깜빡하고 backend브랜치에서 작업했다가 이렇게 됐네요.... 다음부터는 이런 실수 하지 않도록 주의하겠습니다!

Malloc72P committed:
feat : Jwt토큰을 체크하는 인터셉터 구현

- 현재 인터셉터를 적용하지 않아서, 추후에 이를 적용하는 구현이 필요함

Malloc72P committed:
feat : 인터셉터 적용

- 특정 컨트롤러에만 인증로직을 적용하기 위해서 어노테이션 정의하고 사용함 = Auth 어노테이션
- 인터셉터에서 해당 어노테이션이 있는 경우에만 토큰검증을 하도록 구현

Malloc72P committed:
feat : 토큰관련 예외를 처리하는 메서드 추가

Malloc72P committed:
reformat: IntelliJ 자동 리포매팅 수행

Malloc72P committed:
Merge branch 'team-5' into backend

# Conflicts:
#	.gitignore

Malloc72P committed:
merge : git ignore 선 적용 후 team-5의 내용 커밋

Malloc72P commented:
디온! 꼼꼼한 리뷰 정말정말 감사드려요 ㅠㅠ 저희가 실수해서 사실상 PR 2개를 한꺼번에 날린셈이 되어서 너무 고생시킨 것 같아서 죄송한 마음이 드네요 ㅠㅠ
일단 컨플릭은 해결했습니다! 저도 Merge권한이 있어서, 제가 대신 Merge하겠습니다! 
추가로 코멘트 달아주신 내용은 다음 PR을 통해 반영해서 올리겠습니다!
